### PR TITLE
Support customizing engine pool size via user defaults

### PR DESCRIPTION
**Pull request by bowenliang123**
# :mag: Description
## Issue References 🔗

As described.

## Describe Your Solution 🔧

Support customizing `kyuubi.engine.pool.size` with user defaults configs, for the parameter controlling the upper limit of the engine pool for each user.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklists
## 📝 Author Self Checklist

- [x] My code follows the [style guidelines](https://kyuubi.readthedocs.io/en/master/contributing/code/style.html) of this project
- [x] I have performed a self-review
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

## 📝 Committer Pre-Merge Checklist

- [ ] Pull request title is okay.
- [ ] No license issues.
- [ ] Milestone correctly set?
- [ ] Test coverage is ok
- [ ] Assignees are selected.
- [ ] Minimum number of approvals
- [ ] No changes are requested

**Be nice. Be informative.**

### CODE REVIEW

**Comment by bowenliang123** [comment]
cc @pan3793 @yaooqinn

**Comment by yaooqinn** [comment]
Consider using the session conf advisor interface in your prod

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 1]
```diff
@@ -74,7 +74,16 @@ private[kyuubi] class EngineRef(
   private val poolThreshold: Int = Option(KyuubiServer.kyuubiServer).map(_.getConf)
     .getOrElse(KyuubiConf()).get(ENGINE_POOL_SIZE_THRESHOLD)
 
-  private val clientPoolSize: Int = conf.get(ENGINE_POOL_SIZE)
+  private val clientPoolSize: Int = shareLevel match {
+    case ShareLevel.USER =>
+      val serverConf = Option(KyuubiServer.kyuubiServer).map(_.getConf).getOrElse(KyuubiConf())
+      val userPoolSize = serverConf.getAllUserDefaults.get(s"___${user}___${ENGINE_POOL_SIZE.key}")
```
Comment (wForget) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:80]: user default confs seems to have been set to session conf

**Comment by bowenliang123** [comment]
> Consider using the session conf advisor interface in your prod

User defaults configs can be refresh via cli tool, which is more handy than custom conf advisor.
It significantly increase the resource management.

**Comment by bowenliang123** [comment]
Closing this PR, as `kyuubi.engine.pool.size` already supported via user defaults without changes.
Will add proper unit tests for pool size in the future.

**Comment by bowenliang123** [comment]
Closing this PR, as `kyuubi.engine.pool.size` already supported via user defaults without changes.
Will add proper unit tests for pool size in the future.

**Pull request closed without merging**